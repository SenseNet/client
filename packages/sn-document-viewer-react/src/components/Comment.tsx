import Avatar from '@material-ui/core/Avatar'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import CardHeader from '@material-ui/core/CardHeader'
import Collapse from '@material-ui/core/Collapse'
import React, { useState } from 'react'
import { FunctionComponent } from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'
import { Comment as CommentType } from '../models/Comment'
import { componentType } from '../services'
import { deleteComment, RootReducerType } from '../store'

const mapStateToProps = (state: RootReducerType) => ({
  localization: state.sensenetDocumentViewer.localization,
})

const mapDispatchToProps = {
  deleteComment,
}

const StyledCardContent = styled(CardContent)`
  font-size: 14px;
  letter-spacing: 0.2px;
  line-height: 20px;
  word-wrap: break-word;
`
const DeleteButton = (props: CommentPropType) => (
  <Button size="small" onClick={() => props.deleteComment(props.id)}>
    {props.localization.delete || 'delete'}
  </Button>
)

/**
 * Comment prop type
 */
export type CommentPropType = componentType<typeof mapStateToProps, typeof mapDispatchToProps, CommentType>

const MAX_TEXT_LENGTH = 160

/**
 * Represents a single comment component.
 */
export const CommentComponent: FunctionComponent<CommentPropType> = props => {
  const isLongText = props.text && props.text.length > MAX_TEXT_LENGTH
  const [isOpen, setIsOpen] = useState(!isLongText)
  return (
    <Card>
      <CardHeader
        avatar={<Avatar src={props.createdBy.avatarUrl} alt={props.localization.avatarAlt} />}
        title={props.createdBy.displayName}
      />
      <Collapse in={isOpen} timeout="auto" collapsedHeight={isOpen ? '0px' : '78px'}>
        <StyledCardContent>{props.text}</StyledCardContent>
      </Collapse>
      <CardActions>
        {isLongText ? (
          <>
            <Button size="small" onClick={() => setIsOpen(!isOpen)}>
              {isOpen ? props.localization.showLess || 'Show less' : props.localization.showMore || 'Show more'}
            </Button>
            {isOpen ? <DeleteButton {...props} /> : null}
          </>
        ) : (
          <DeleteButton {...props} />
        )}
      </CardActions>
    </Card>
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CommentComponent)
