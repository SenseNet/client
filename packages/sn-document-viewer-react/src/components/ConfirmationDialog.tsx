import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'
import React from 'react'
import { FunctionComponent } from 'react'

/**
 * Interface for confirmation dialog component properties
 */
export interface ConfirmationDialogProps {
  dialogTitle: string
  okButtonText?: string
  cancelButtonText?: string
  isOpen: boolean
  onClose: (isCanceled: boolean) => void
}

/**
 * Represents a confirmation dialog
 */
export const ConfirmationDialog: FunctionComponent<ConfirmationDialogProps> = props => {
  const handleCancel = () => props.onClose(true)
  const handleOk = () => props.onClose(false)
  return (
    <Dialog
      disableBackdropClick={true}
      disableEscapeKeyDown={true}
      maxWidth="md"
      aria-labelledby="confirmation-dialog-title"
      aria-describedby="confirmation-dialog-description"
      open={props.isOpen}>
      <DialogTitle id="confirmation-dialog-title">{props.dialogTitle}</DialogTitle>
      <DialogContent>
        <DialogContentText
          dangerouslySetInnerHTML={{ __html: props.children as string }}
          id="confirmation-dialog-description"
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleOk} variant="contained" color="primary">
          {props.okButtonText || 'ok'}
        </Button>
        <Button onClick={handleCancel}>{props.cancelButtonText || 'cancel'}</Button>
      </DialogActions>
    </Dialog>
  )
}
