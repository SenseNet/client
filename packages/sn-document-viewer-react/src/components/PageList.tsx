import Grid from '@material-ui/core/Grid'
import React, { useCallback, useEffect, useRef, useState } from 'react'
import { debounce } from '@sensenet/client-utils'
import { PreviewImageData } from '@sensenet/client-core'
import { Dimensions, ImageUtil } from '../services'
import { ZoomMode } from '../models/viewer-state'
import { usePreviewImages, useViewerState } from '../hooks'
import { CommentsContextProvider } from '../context/comments'
import { Page } from './'

/**
 * Defines the own properties for the PageList component
 */
export interface PageListProps {
  tolerance: number
  padding: number
  id: string
  elementName: string
  zoomMode: ZoomMode
  zoomLevel: number
  fitRelativeZoomLevel: number
  images: 'preview' | 'thumbnail'
  activePage?: number
  onPageClick: (ev: React.MouseEvent<HTMLElement>, pageIndex: number) => void
  style?: React.CSSProperties
  showWidgets: boolean
}

export const PageList: React.FC<PageListProps> = (props) => {
  const [marginTop, setMarginTop] = useState(0)
  const [marginBottom, setMarginBottom] = useState(0)
  const [visiblePages, setVisiblePages] = useState<PreviewImageData[]>([])
  const [scrollState, setScrollState] = useState(0)
  const viewportElement = useRef<HTMLElement>()
  const viewerState = useViewerState()
  const [resizeToken, setResizeToken] = useState(0)
  const [viewport, setViewport] = useState<Dimensions>({ width: 0, height: 0 })

  const pages = usePreviewImages()

  const requestResize = useCallback(
    debounce(() => {
      setResizeToken(Math.random())
    }, 100),
    [],
  )

  const onScroll = useCallback(() => {
    requestResize()
    setScrollState((viewportElement.current && viewportElement.current.scrollTop) || 0)
  }, [requestResize])

  useEffect(() => {
    const currentViewport = viewportElement.current
    if (currentViewport) {
      currentViewport.addEventListener('scroll', onScroll)
    }
    return () => currentViewport && currentViewport.removeEventListener('scroll', onScroll)
  }, [onScroll, viewportElement])

  useEffect(() => {
    window.addEventListener('resize', requestResize)
    return () => window.removeEventListener('resize', requestResize)
  }, [requestResize])

  useEffect(() => {
    if (viewportElement && viewportElement.current) {
      setViewport({
        height: viewportElement.current.clientHeight - props.padding * 2,
        width: viewportElement.current.clientWidth - props.padding * 2,
      })
    }
  }, [props.padding, resizeToken, viewportElement])

  useEffect(() => {
    if (!pages.imageData.length) {
      return
    }

    let defaultWidth!: number
    let defaultHeight!: number

    const _visiblePages = pages.imageData.map((p) => {
      if ((p && !defaultWidth) || !defaultHeight) {
        ;[defaultWidth, defaultHeight] = [p.Width, p.Height]
      }

      if (!p.Width || !p.Height) {
        ;[p.Width, p.Height] = [defaultWidth, defaultHeight]
      }

      const relativeSize = ImageUtil.getImageSize(
        {
          width: viewport.width,
          height: viewport.height,
        },
        {
          width: p.Width,
          height: p.Height,
          rotation: (p.Attributes && p.Attributes.degree) || 0,
        },
        props.zoomMode,
        props.zoomLevel,
        props.fitRelativeZoomLevel,
      )

      return {
        ...p,
        Width: relativeSize.width,
        Height: relativeSize.height,
      }
    })

    let _marginTop = 0
    let _pagesToSkip = 0

    while (
      _visiblePages[_pagesToSkip] &&
      _marginTop + _visiblePages[_pagesToSkip].Height + props.tolerance < scrollState
    ) {
      _marginTop += _visiblePages[_pagesToSkip].Height + props.padding * 2
      _pagesToSkip++
    }

    let _pagesToTake = 1
    let _pagesHeight = 0

    while (_visiblePages[_pagesToSkip + _pagesToTake] && _pagesHeight < viewport.height + props.tolerance) {
      _pagesHeight += _visiblePages[_pagesToSkip + _pagesToTake].Height + props.padding * 2
      _pagesToTake++
    }

    let _marginBottom = 0
    for (let i = _pagesToSkip + _pagesToTake - 1; i < _visiblePages.length - 1; i++) {
      _marginBottom += _visiblePages[i].Height + props.padding * 2
    }

    setMarginTop(_marginTop)
    setMarginBottom(_marginBottom)
    const newVisiblePages = _visiblePages.slice(_pagesToSkip, _pagesToSkip + _pagesToTake)
    setVisiblePages(newVisiblePages)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    pages.imageData,
    props.fitRelativeZoomLevel,
    props.padding,
    props.tolerance,
    props.zoomLevel,
    props.zoomMode,
    scrollState,
    viewerState.activePages,
    viewport.height,
    viewport.width,
  ])

  const updateScrollState = useCallback(
    debounce(() => {
      if (props.images === 'preview')
        viewerState.updateState({ activePages: [(visiblePages[0] && visiblePages[0].Index) || 1] })
    }, 100),
    [viewerState, visiblePages],
  )

  useEffect(updateScrollState, [scrollState])

  return (
    <Grid
      item={true}
      style={{ ...props.style, flexGrow: 1, flexShrink: 1, overflow: 'auto', height: '100%' }}
      id={props.id}
      innerRef={viewportElement}>
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'center',
          alignItems: 'center',
          paddingTop: marginTop || 0,
          paddingBottom: marginBottom || 0,
        }}>
        {visiblePages.map((page) => (
          <CommentsContextProvider page={page.Index} key={page.Index} images={props.images}>
            <Page
              showWidgets={props.showWidgets}
              viewportWidth={viewport.width}
              viewportHeight={viewport.height}
              imageIndex={page.Index}
              onClick={(ev) => props.onPageClick(ev, page.Index)}
              zoomMode={props.zoomMode}
              zoomLevel={props.zoomLevel}
              fitRelativeZoomLevel={props.fitRelativeZoomLevel}
              elementName={props.elementName}
              image={props.images}
              margin={props.padding}
            />
          </CommentsContextProvider>
        ))}
      </div>
    </Grid>
  )
}
