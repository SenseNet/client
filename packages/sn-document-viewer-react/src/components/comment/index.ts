export * from './comment'
export * from './create-comment'
export * from './delete-comment-button'
export * from './style'
