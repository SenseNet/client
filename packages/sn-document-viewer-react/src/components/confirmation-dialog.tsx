import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle } from '@material-ui/core'
import React from 'react'

/**
 * Interface for confirmation dialog component properties
 */
export interface ConfirmationDialogProps {
  dialogTitle: string
  okButtonText?: string
  cancelButtonText?: string
  isOpen: boolean
  onClose: (isCanceled: boolean) => void
}

/**
 * Represents a confirmation dialog
 */
export const ConfirmationDialog: React.FC<ConfirmationDialogProps> = (props) => {
  const handleCancel = () => props.onClose(true)
  const handleOk = () => props.onClose(false)
  return (
    <Dialog
      disableBackdropClick={true}
      disableEscapeKeyDown={true}
      maxWidth="md"
      aria-labelledby="confirmation-dialog-title"
      aria-describedby="confirmation-dialog-description"
      open={props.isOpen}>
      <DialogTitle id="confirmation-dialog-title">{props.dialogTitle}</DialogTitle>
      <DialogContent>
        <DialogContentText
          dangerouslySetInnerHTML={{ __html: props.children as string }}
          id="confirmation-dialog-description"
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleOk} variant="contained" color="primary">
          {props.okButtonText || 'ok'}
        </Button>
        <Button onClick={handleCancel}>{props.cancelButtonText || 'cancel'}</Button>
      </DialogActions>
    </Dialog>
  )
}
