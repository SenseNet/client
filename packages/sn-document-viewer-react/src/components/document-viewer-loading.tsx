import { createStyles, makeStyles } from '@material-ui/core'
import Typography from '@material-ui/core/Typography'
import React from 'react'
import { useLocalization } from '../hooks'

const useStyles = makeStyles(() => {
  return createStyles({
    mainWrapper: {
      display: 'flex',
      justifyContent: 'space-around',
      alignContent: 'center',
      alignItems: 'center',
      height: '100%',
    },
    loaderWrapper: {
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      flexDirection: 'column',
      maxWidth: 500,
      margin: '.5em 0 .6em 0',
    },
    typography: {
      marginTop: '1rem',
      fontWeight: 'bolder',
    },
  })
})

interface DocumentViewerLoadingProps {
  image: string
}

export const DocumentViewerLoading: React.FC<DocumentViewerLoadingProps> = (props) => {
  const classes = useStyles()
  const localization = useLocalization()

  return (
    <div className={classes.mainWrapper}>
      <div className={classes.loaderWrapper}>
        <img src={props.image} alt="Loader" />
        <Typography variant="h5" color="textSecondary" align="center" className={classes.typography}>
          {localization.loadingDocument}
        </Typography>
      </div>
    </div>
  )
}
