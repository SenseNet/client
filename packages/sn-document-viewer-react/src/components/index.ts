export * from './comment'
export * from './document-widgets'
export * from './page-widgets'
export * from './ConfirmationDialog'
export * from './DocumentViewer'
export * from './DocumentViewerError'
export * from './DocumentViewerLayout'
export * from './DocumentViewerLoading'
export * from './DocumentViewerRegeneratePreviews'
export * from './LayoutAppBar'
export * from './Page'
export * from './PageList'
export * from './styles'
