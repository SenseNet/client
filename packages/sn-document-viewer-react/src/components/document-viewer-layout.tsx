import Drawer from '@material-ui/core/Drawer'
import createStyles from '@material-ui/core/styles/createStyles'
import makeStyles from '@material-ui/core/styles/makeStyles'
import Typography from '@material-ui/core/Typography'
import React, { useCallback, useEffect } from 'react'
import { CommentsContext } from '../context/comments'
import { useLocalization, useViewerSettings, useViewerState } from '../hooks'
import { Comment, CreateComment } from './comment'
import { PageList, Thumbnails } from './'

export const THUMBNAIL_PADDING = 16
export const THUMBNAIL_NAME = 'Thumbnail'
export const THUMBNAIL_CONTAINER_ID = 'sn-document-viewer-thumbnails'
export const PAGE_PADDING = 8
export const PAGE_NAME = 'Page'
export const PAGE_CONTAINER_ID = 'sn-document-viewer-pages'

const useStyles = makeStyles(() => {
  return createStyles({
    mainWrapper: {
      width: '100%',
      height: '100%',
      display: 'flex',
      flexDirection: 'column',
      flexGrow: 1,
    },
    layout: {
      display: 'flex',
      flexGrow: 1,
      width: '100%',
      overflow: 'hidden',
      zIndex: 0,
      position: 'relative',
    },
    commentContainer: {
      display: 'flex',
      flexFlow: 'column',
      padding: '15px',
      justifyContent: 'space-between',
      alignItems: 'center',
      flexWrap: 'wrap',
      overflow: 'auto',
    },
  })
})

interface ScrollToOptions {
  containerId: string
  index: number
  itemName: string
  padding: number
  smoothScroll: boolean
}

export const DocumentViewerLayout: React.FC = () => {
  const classes = useStyles()
  const viewerState = useViewerState()
  const localization = useLocalization()
  const viewerSettings = useViewerSettings()

  const scrollToImage = useCallback(({ containerId, index, itemName, padding, smoothScroll }: ScrollToOptions) => {
    const container = document.getElementById(containerId)
    const item = document.querySelector(`.${itemName}`)
    if (container && container.scrollTo && item) {
      container.scrollTo({
        top: (item.clientHeight + padding * 4) * (index - 1),
        behavior: smoothScroll ? 'smooth' : 'auto',
      })
    }
  }, [])

  const scrollTo = useCallback(
    (index: number, smoothScroll = true) => {
      scrollToImage({
        containerId: PAGE_CONTAINER_ID,
        itemName: PAGE_NAME,
        padding: PAGE_PADDING,
        index,
        smoothScroll,
      })

      scrollToImage({
        containerId: THUMBNAIL_CONTAINER_ID,
        itemName: THUMBNAIL_NAME,
        padding: THUMBNAIL_PADDING,
        index,
        smoothScroll,
      })
    },
    [scrollToImage],
  )

  useEffect(() => {
    const observer = viewerState.pageToGo.subscribe(({ page }) => {
      scrollTo(page)
    })
    return () => observer.dispose()
  }, [scrollTo, viewerState.pageToGo])

  return (
    <div className={classes.mainWrapper}>
      {viewerSettings.renderAppBar()}
      <div className={classes.layout}>
        <Drawer
          variant={'persistent'}
          open={viewerState.showThumbnails}
          anchor="left"
          SlideProps={viewerSettings.drawerSlideProps}
          PaperProps={{
            style: {
              position: 'relative',
              width: viewerState.showThumbnails ? '200px' : 0,
              height: '100%',
              overflow: 'hidden',
            },
          }}>
          {viewerState.showThumbnails && <Thumbnails onPageClick={(index) => scrollTo(index)} />}
        </Drawer>
        <PageList onPageClick={(index) => scrollTo(index)} />
        <Drawer
          variant={'persistent'}
          open={viewerState.showComments}
          anchor="right"
          SlideProps={viewerSettings.drawerSlideProps}
          PaperProps={{
            style: {
              position: 'relative',
              width: viewerState.showComments ? '340px' : 0,
              height: '100%',
              overflow: 'hidden',
            },
          }}>
          <div className={classes.commentContainer}>
            <Typography variant="h4">{localization.commentSideBarTitle}</Typography>
            <CreateComment />
            <CommentsContext.Consumer>
              {(commentsContext) =>
                commentsContext.comments.map((comment) => <Comment key={comment.id} comment={comment} />)
              }
            </CommentsContext.Consumer>
          </div>
        </Drawer>
      </div>
    </div>
  )
}
