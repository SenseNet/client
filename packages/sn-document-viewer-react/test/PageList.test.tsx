import { shallow } from 'enzyme'
import React from 'react'
import { PageListComponent } from '../src/components/PageList'

describe('PageList component', () => {
  it('should match snapshot with one page', () => {
    const wrapper = shallow(
      <PageListComponent
        pages={[{ Height: 100, Width: 50, Index: 1 }]}
        showWidgets={true}
        id="sn-document-viewer-pages"
        zoomMode="originalSize"
        zoomLevel={1}
        fitRelativeZoomLevel={1}
        onPageClick={() => jest.fn}
        elementNamePrefix="Page-"
        images="preview"
        tolerance={0}
        padding={8}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })
})
