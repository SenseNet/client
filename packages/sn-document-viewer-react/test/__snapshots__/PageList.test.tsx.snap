// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageList component should match snapshot with one page 1`] = `
<WithStyles(Grid)
  id="sn-document-viewer-pages"
  item={true}
  style={
    Object {
      "flexGrow": 1,
      "flexShrink": 1,
      "height": "100%",
      "overflow": "auto",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingBottom": 0,
        "paddingTop": 0,
      }
    }
  >
    <Connect(Page)
      elementNamePrefix="Page-"
      fitRelativeZoomLevel={1}
      image="preview"
      imageIndex={1}
      key="1"
      margin={8}
      onClick={[Function]}
      showWidgets={true}
      viewportHeight={110}
      viewportWidth={110}
      zoomLevel={1}
      zoomMode="originalSize"
    />
  </div>
</WithStyles(Grid)>
`;
